File: framec_tests/java/StateContextStack/StateContextStackTest.java
package framec_tests.java.StateContextStack;

import static org.junit.Assert.assertEquals;

import java.util.ArrayList;
import java.util.Arrays;

import org.junit.Test;

class StateContextStackController extends StateContextStack {

    StateContextStackController() {
      super();
    }

protected void log_do(String msg) {
    this.tape.add(msg);
}
}

public class StateContextStackTest {
   // Test that a pop restores a pushed state.
    @Test
    public void TestPushPop(){
        StateContextStackController sm = new StateContextStackController();
        assertEquals( "0", sm.state_info());
        sm.push();
        sm.to_b();
        assertEquals("1", sm.state_info());
        sm.pop();
        assertEquals("0", sm.state_info());
    }

    /* Test that multiple states can be pushed and subsequently restored by pops, LIFO style. */
    @Test
    public void TestMultiplePushPops(){
        StateContextStackController sm = new StateContextStackController();
        assertEquals("0", sm.state_info());
        sm.push();
        sm.to_c();
        sm.push();
        sm.to_a();
        sm.push();
        sm.push();
        sm.to_c(); // no push
        sm.to_b();
        sm.push();
        sm.to_c();
        sm.push(); // stack top-to-bottom: C, B, A, A, C, A
        sm.to_a();
        assertEquals("0", sm.state_info());
        sm.pop();
        assertEquals("2", sm.state_info());
        sm.to_a();
        assertEquals("0", sm.state_info());
        sm.pop();
        assertEquals("1", sm.state_info());
        sm.pop();
        assertEquals("0", sm.state_info());
        sm.pop();
        assertEquals("0", sm.state_info());
        sm.pop();
        assertEquals("2", sm.state_info());
        sm.to_b();
        sm.push();
        sm.to_c();
        sm.push(); // stack top-to-bottom: C, B, A
        sm.to_a();
        sm.to_b();
        assertEquals("1", sm.state_info());
        sm.pop();
        assertEquals("2", sm.state_info());
        sm.pop();
        assertEquals("1", sm.state_info());
        sm.pop();
        assertEquals("0", sm.state_info());

    }

    /*  Test that pop transitions trigger enter/exit events. */
    @Test
    public void TestPopTransitionEvents(){
        StateContextStackController sm = new StateContextStackController();
        sm.to_b();
        sm.push();
        sm.to_a();
        sm.push();
        sm.to_c();
        sm.push(); // stack top-to-bottom: C, A, B
        sm.to_a();
        sm.tape = new ArrayList<>();
        assertEquals("0", sm.state_info());
        sm.pop();
        assertEquals("2", sm.state_info());
        assertEquals(sm.tape, Arrays.asList("A:<", "C:>"));
        sm.tape = new ArrayList<>();
        sm.pop();
        sm.pop();
        assertEquals("1", sm.state_info());
        assertEquals(sm.tape, Arrays.asList("C:<", "A:>", "A:<", "B:>"));
  
    }
    /* Test that pop change-states do not trigger enter/exit events. */
    @Test
    public void TestPopChangesStateNoEvents(){
        StateContextStackController sm = new StateContextStackController();
        sm.to_b();
        sm.push();
        sm.to_a();
        sm.push();
        sm.to_c();
        sm.push(); // stack top-to-bottom: C, A, B
        sm.to_a();
        sm.tape = new ArrayList<>();
        assertEquals("0", sm.state_info());
        sm.pop_change();
        assertEquals("2", sm.state_info());
        assertEquals(0, sm.tape.size());
        sm.pop();
        sm.pop_change();
        assertEquals("1", sm.state_info());
        assertEquals(Arrays.asList("C:<", "A:>"), sm.tape);
    }
    /* Test that state variables are restored after pop. */
    @Test
    public void TestPopRestoresStateVariables(){
        StateContextStackController sm = new StateContextStackController();
        sm.inc();
        sm.inc();
        sm.push();
        assertEquals("0", sm.state_info());
        assertEquals(2, sm.value());
        sm.to_b();
        sm.inc();
        sm.push();
        assertEquals("1", sm.state_info());
        assertEquals(5, sm.value());
        sm.to_c();
        sm.inc();
        sm.inc();
        sm.inc();
        sm.push();
        assertEquals("2", sm.state_info());
        assertEquals(sm.value(), 30);
        sm.to_a();
        sm.inc();
        assertEquals("0", sm.state_info());
        assertEquals(1, sm.value());
        sm.pop();
        assertEquals("2", sm.state_info());
        assertEquals(30, sm.value());
        sm.pop();
        assertEquals("1", sm.state_info());
        assertEquals(5, sm.value());
        sm.to_a();
        sm.inc();
        sm.inc();
        sm.inc();
        sm.push();
        assertEquals("0", sm.state_info());
        assertEquals(3, sm.value());
        sm.to_c();
        sm.inc();
        assertEquals("2", sm.state_info());
        assertEquals(10, sm.value());
        sm.pop();
        assertEquals("0", sm.state_info());
        assertEquals(3, sm.value());
        sm.pop();
        assertEquals("0", sm.state_info());
        assertEquals(2, sm.value());
    }
    @Test
    public void TestPushStoresStateVariableSnapshot(){
        StateContextStackController sm = new StateContextStackController();
        sm.inc();
        sm.inc();
        sm.push();
        assertEquals("0", sm.state_info());
        assertEquals(2, sm.value());
        sm.inc();
        sm.inc();
        assertEquals(4, sm.value());

        sm.to_b();
        sm.inc();
        sm.push();
        assertEquals("1", sm.state_info());
        assertEquals(5, sm.value());
        sm.inc();
        sm.inc();
        assertEquals(15, sm.value()); // these changes should be forgotten

        sm.to_c();
        sm.inc();
        sm.inc();
        sm.inc();
        sm.push();
        assertEquals("2", sm.state_info());
        assertEquals(30, sm.value());
        sm.inc();
        assertEquals(40, sm.value()); // forgotten

        sm.to_a();
        sm.inc();
        assertEquals("0", sm.state_info());
        assertEquals(1, sm.value());

        sm.pop();
        assertEquals("2", sm.state_info());
        assertEquals(30, sm.value());

        sm.pop();
        assertEquals("1", sm.state_info());
        assertEquals(5, sm.value());

        sm.to_a();
        sm.inc();
        sm.inc();
        sm.inc();
        sm.push();
        assertEquals("0", sm.state_info());
        assertEquals(3, sm.value());
        sm.inc();
        assertEquals(4, sm.value()); //forgotten

        sm.to_c();
        sm.inc();
        assertEquals("2", sm.state_info());
        assertEquals(10, sm.value());

        sm.pop();
        assertEquals("0", sm.state_info());
        assertEquals(3, sm.value());

        sm.pop();
        assertEquals("0", sm.state_info());
        assertEquals(2, sm.value());
    }
    
}

File: framec_tests/csharp/StateContextStack/StateContextStackTest.cs
using System.IO;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using NUnit.Framework;

namespace StateContextStack
{
    class StateContextStackController : StateContextStack
    {
        public StateContextStackController() : base()
        {
        }
        public void log_do(string msg)
        {
            this.tape.Add(msg);
        }
    }
    [TestFixture]
    public class StateContextStackTest
    {
        [Test]
        public void TestPushPop()
        {
            StateContextStackController sm = new StateContextStackController();
            Assert.AreEqual("0", sm.state_info());
            sm.push();
            sm.to_b();
            Assert.AreEqual("1", sm.state_info());
            sm.pop();
            Assert.AreEqual("0", sm.state_info());
        }

        /* Test that multiple states can be pushed and subsequently restored by pops, LIFO style. */
        [Test]
        public void TestMultiplePushPops()
        {
            StateContextStackController sm = new StateContextStackController();
            Assert.AreEqual("0", sm.state_info());
            sm.push();
            sm.to_c();
            sm.push();
            sm.to_a();
            sm.push();
            sm.push();
            sm.to_c(); // no push
            sm.to_b();
            sm.push();
            sm.to_c();
            sm.push(); // stack top-to-bottom: C, B, A, A, C, A
            sm.to_a();
            Assert.AreEqual("0", sm.state_info());
            sm.pop();
            Assert.AreEqual("2", sm.state_info());
            sm.to_a();
            Assert.AreEqual("0", sm.state_info());
            sm.pop();
            Assert.AreEqual("1", sm.state_info());
            sm.pop();
            Assert.AreEqual("0", sm.state_info());
            sm.pop();
            Assert.AreEqual("0", sm.state_info());
            sm.pop();
            Assert.AreEqual("2", sm.state_info());
            sm.to_b();
            sm.push();
            sm.to_c();
            sm.push(); // stack top-to-bottom: C, B, A
            sm.to_a();
            sm.to_b();
            Assert.AreEqual("1", sm.state_info());
            sm.pop();
            Assert.AreEqual("2", sm.state_info());
            sm.pop();
            Assert.AreEqual("1", sm.state_info());
            sm.pop();
            Assert.AreEqual("0", sm.state_info());
        }

        /* Test that pop transitions trigger enter/exit events. */
        [Test]
        public void TestPopTransitionEvents()
        {
            StateContextStackController sm = new StateContextStackController();
            sm.to_b();
            sm.push();
            sm.to_a();
            sm.push();
            sm.to_c();
            sm.push(); // stack top-to-bottom: C, A, B
            sm.to_a();
            sm.tape = new List<string>();
            Assert.AreEqual("0", sm.state_info());
            sm.pop();
            Assert.AreEqual("2", sm.state_info());
            Assert.AreEqual(new List<string> { "A:<", "C:>" }, sm.tape);
            sm.tape = new List<string>();
            sm.pop();
            sm.pop();
            Assert.AreEqual("1", sm.state_info());
            Assert.AreEqual(new List<string> { "C:<", "A:>", "A:<", "B:>" }, sm.tape);
        }

        /* Test that pop change-states do not trigger enter/exit events. */
        [Test]
        public void TestPopChangesStateNoEvents()
        {
            StateContextStackController sm = new StateContextStackController();
            sm.to_b();
            sm.push();
            sm.to_a();
            sm.push();
            sm.to_c();
            sm.push(); // stack top-to-bottom: C, A, B
            sm.to_a();
            sm.tape = new List<string>();
            Assert.AreEqual("0", sm.state_info());
            sm.pop_change();
            Assert.AreEqual("2", sm.state_info());
            Assert.AreEqual(0, sm.tape.Count);
            sm.pop();
            sm.pop_change();
            Assert.AreEqual("1", sm.state_info());
            Assert.AreEqual(new List<string> { "C:<", "A:>" }, sm.tape);
        }

        [Test]
        public void TestPopRestoresStateVariables()
        {
            StateContextStackController sm = new StateContextStackController();
            sm.inc();
            sm.inc();
            sm.push();
            Assert.AreEqual("0", sm.state_info());
            Assert.AreEqual(2, sm.value());
            sm.to_b();
            sm.inc();
            sm.push();
            Assert.AreEqual("1", sm.state_info());
            Assert.AreEqual(5, sm.value());
            sm.to_c();
            sm.inc();
            sm.inc();
            sm.inc();
            sm.push();
            Assert.AreEqual("2", sm.state_info());
            Assert.AreEqual(30, sm.value());
            sm.to_a();
            sm.inc();
            Assert.AreEqual("0", sm.state_info());
            Assert.AreEqual(1, sm.value());
            sm.pop();
            Assert.AreEqual("2", sm.state_info());
            Assert.AreEqual(30, sm.value());
            sm.pop();
            Assert.AreEqual("1", sm.state_info());
            Assert.AreEqual(5, sm.value());
            sm.to_a();
            sm.inc();
            sm.inc();
            sm.inc();
            sm.push();
            Assert.AreEqual("0", sm.state_info());
            Assert.AreEqual(3, sm.value());
            sm.to_c();
            sm.inc();
            Assert.AreEqual("2", sm.state_info());
            Assert.AreEqual(10, sm.value());
            sm.pop();
            Assert.AreEqual("0", sm.state_info());
            Assert.AreEqual(3, sm.value());
            sm.pop();
            Assert.AreEqual("0", sm.state_info());
            Assert.AreEqual(2, sm.value());
        }

        [Test]
        public void TestPushStoresStateVariableSnapshot()
        {
            StateContextStackController sm = new StateContextStackController();
            sm.inc();
            sm.inc();
            sm.push();
            Assert.AreEqual("0", sm.state_info());
            Assert.AreEqual(2, sm.value());
            sm.inc();
            sm.inc();
            Assert.AreEqual(4, sm.value());

            sm.to_b();
            sm.inc();
            sm.push();
            Assert.AreEqual("1", sm.state_info());
            Assert.AreEqual(5, sm.value());
            sm.inc();
            sm.inc();
            Assert.AreEqual(15, sm.value()); // these changes should be forgotten

            sm.to_c();
            sm.inc();
            sm.inc();
            sm.inc();
            sm.push();
            Assert.AreEqual("2", sm.state_info());
            Assert.AreEqual(30, sm.value());
            sm.inc();
            Assert.AreEqual(40, sm.value()); // forgotten

            sm.to_a();
            sm.inc();
            Assert.AreEqual("0", sm.state_info());
            Assert.AreEqual(1, sm.value());

            sm.pop();
            Assert.AreEqual("2", sm.state_info());
            Assert.AreEqual(30, sm.value());

            sm.pop();
            Assert.AreEqual("1", sm.state_info());
            Assert.AreEqual(5, sm.value());

            sm.to_a();
            sm.inc();
            sm.inc();
            sm.inc();
            sm.push();
            Assert.AreEqual("0", sm.state_info());
            Assert.AreEqual(3, sm.value());
            sm.inc();
            Assert.AreEqual(4, sm.value()); //forgotten

            sm.to_c();
            sm.inc();
            Assert.AreEqual("2", sm.state_info());
            Assert.AreEqual(10, sm.value());

            sm.pop();
            Assert.AreEqual("0", sm.state_info());
            Assert.AreEqual(3, sm.value());

            sm.pop();
            Assert.AreEqual("0", sm.state_info());
            Assert.AreEqual(2, sm.value());
        }



    }
}